### DISCUSSION TITLE
Is IPv6 supported ?

### DISCUSSION THREAD

**Discussion by jmarcgit** [mikejac/node-red-contrib-google-smarthome#613] [Q&A]
Hi,

I'm asking this because CGNAT is deployed right now. Any experience with that ?

**Comment by Caprico85** [comment]
Hi,

As far as I know, we don't have anything in the source code that explicitly prevents IPv6. Theoretically, it should work. Practically, I'm currently trying to figure it out. I deleted the A record from my dual-stack setup and left only the AAAA record. Since Google now only knows my IPv6 address, it should access it via IPv6. However, it is still using IPv4. I probably need to wait until Google's DNS caches expire. I'll report back when I know more.

**Comment by jmarcgit** [comment]
Thank you. I did some tests on my side. I can reach the check page from inside and outside my network using IPv6. I spent HOURS because the online http clients I tested don't work with IPv6. I will pair my server with Google tomorrow and will report here.

update : it works for me

**Comment by Caprico85** [comment]
After a few days of working with IPv6, I can confirm that it works for me too.

It is noticeably slower than IPv4. But I guess that is something about my network.

**Comment by jmarcgit** [comment]
@Caprico85 I believe I found potential problems related to local fullfilment. Are you sure the local protocol is still working ?

1- In local mode there is a check of the source ipv6 address. And nodered smarthome will only accept local link Ipv6  addresses. Any connection from a local device throught a global IPv6 will be blocked. It's possible to say if 2 IPv6 are in the same network but it requires a calculation (extract prefix of 2 addresses and compare them). I believe this logic also works for IPv4 (only difference is the address length) so we can write a single function that compares 2 addresses and says if they are in the same subnet or not.

2- Looks like the UDP server is using IPv4 (udp4 is mentionned when creating the socket). This means that IPv4 will also be used to query the express webserver. I switched nodered to IPv6 and it looks like there no IPv4 listening socket anymore in nodered. So I believe UDP advertizing is broken. Maybe we have to check if nodered is started with ipv6 binding and then create an udp6 socket instead ?

3- Maybe MDNS advertizing still works but I'm afraid that the HTTP requests could be blocked because of 1- In my experience avahi is advertzing using a global ipv6 whenever possible.

**Reply by jmarcgit** [reply]
![image](https://github.com/user-attachments/assets/c700c99a-a0b9-4ade-9ef1-b0d3a82bc4b8)

Here you see that the local http server is listening on 9998 (tcp6) but the UDP listener (9998 udp) is listening to IPv4 connections only. Therefore local fullfilment will use IPv4. Because on many systems the tcp6 socket allows connections throught IPv4 as well. The easiest way to fix that is: get the web server socket configuration when it's started and apply these settings to the udp socket (socket family and binding address). I can create a pull request.

**Comment by Caprico85** [comment]
> Are you sure the local protocol is still working ?

Local fulfillment via UDP4 discovery was still working for me.

> 1- In local mode there is a check of the source ipv6 address. 

You're right. isLocalIP in HttpActions.js needs to be changed. to check global addresses by comparing the suffix.

We can use ip-range-check for this. We already use this module to do the ipv4 checking. It also supports IPv6 calculations.

```
var ipRangeCheck = require("ip-range-check");

// first address is the sender address
// second address is our local IP address, including the prefix length
console.log(ipRangeCheck("2003:ec:ef16:aaaa:2ecf:67ff:1111:1111", ["2003:ec:ef16:fe00:2ecf:67ff:fec8:229f/64"]))
```

For IPv4 we should keep the current calculation and not take the hosts ip address into account. If running in Docker, we cannot compare the network the network prefixes as docker are running in a docker specific subnet with its own prefix. Of course this excludes all Docker users from using IPv6. But by default Docker doesn't support IPv6 anyway, so thats okay. We should only add a note to the Docker section in the local fulfillment guide that IPv4 needs to be used.

---

> 2- Looks like the UDP server is using IPv4 (udp4 is mentionned when creating the socket). This means that IPv4 will also be used to query the express webserver. I switched nodered to IPv6 and it looks like there no IPv4 listening socket anymore in nodered. So I believe UDP advertizing is broken. Maybe we have to check if nodered is started with ipv6 binding and then create an udp6 socket instead ?

Correct. Currently, UDP only listens for IPv4. I added UDP6 support in d777a9686fb36f76e66db876da339a0f32220cb6 (not yet released as a new version). I don't check if Node-RED uses IPv4 or IPv6, I simply start sockets for UDP4 and UDP6 anyway.

So far, I was not able to receive any UDP packets from the smart speaker via IPv6. I tried by setting the Broadcast address to my host's IPv6 address and to `ff02::1` (which seems to be the equivalent multicast address to 255.255.255.255). I get no packets, neither via Node-RED nor using `netcat -ul 1234`. I'm not sure if I am using the wrong addresses or if the smart speakers doesn't support UDP6 discovery at all. Unfortunately I cannot find any information on the internet regarding local fulfillment via IPv6.

It shouldn't be a problem that the IP range check does not yet support IPv6. HTTP communication with it's ip check only happens after UDP discovery is done.

---

> 3- Maybe MDNS advertizing still works but I'm afraid that the HTTP requests could be blocked because of 1- In my experience avahi is advertzing using a global ipv6 whenever possible.

So far, I wasn't able to get MDNS discovery working, not even with IPv4. I am still trying to investigate, but I'll focus on UDP first.
